racket/collects/racket/match/parse.rkt

```diff
@@ -52,6 +55,131 @@
           (rest suffix)
           (if (eq? ddk-size #t) 0 ddk-size)))
 
+
+;; gen-closed-mode :: syntax?
+;;                    (listof identifier?)
+;;                    (listof identifier?)
+;;                    -> (listof syntax?)
+(define (gen-closed-mode mode k-ids ref-ids)
+  (cond
+    [(eq? mode #t)
+     (list #`(λ ()
+               (define seen (hash-copy-clear e #:kind 'mutable))
+               (define cnt
+                 (+ #,@(for/list ([k-id (in-list k-ids)]
+                                  [ref-id (in-list ref-ids)])
+                         #`(cond
+                             [(or (hash-has-key? seen #,k-id)
+                                  (user-def? #,ref-id))
+                              0]
+                             [else
+                              (hash-set! seen #,k-id #t)
+                              1]))))
+               (= (hash-count e) cnt)))]
+    [else '()]))
+
+;; do-hash :: syntax? stx-list? (or/c #t #f syntax?) -> syntax?
+(define (do-hash stx kvps mode)
+  (define kvp-list (syntax->list kvps))
+  (define-values (k-exprs v-pats def-exprs def-ids)
+    (for/fold ([k-exprs '()]
+               [v-pats '()]
+               [def-exprs '()]
+               [def-ids '()]
+               #:result (values (reverse k-exprs)
+                                (reverse v-pats)
+                                (reverse def-exprs)
+                                (reverse def-ids)))
+              ([kvp (in-list kvp-list)])
+      (syntax-case kvp ()
+        [(k-expr v-pat #:default def-expr)
+         (values (cons #'k-expr k-exprs)
+                 (cons #'v-pat v-pats)
+                 (cons #'def-expr def-exprs)
+                 (cons #'user-def def-ids))]
+        [(k-expr v-pat)
+         (values (cons #'k-expr k-exprs)
+                 (cons #'v-pat v-pats)
+                 (cons #'undef def-exprs)
+                 (cons #'undef def-ids))]
+        [_ (raise-syntax-error #f "expect a key-value group" stx kvp)])))
+
+  (define k-ids (generate-temporaries k-exprs))
+  (define ref-ids (generate-temporaries k-exprs))
+
+  (with-syntax ([(k-id ...) k-ids]
+                [(ref-id ...) ref-ids]
+                [(k-expr ...) k-exprs]
+                [(v-pat ...) v-pats]
+                [(def-expr ...) def-exprs]
+                [(def-id ...) def-ids])
+    (parse
```

Comment:
Is it necessary to construct a pattern and then parse it? I would prefer to go directly to the relevant AST (or build new kinds of AST if that's needed).

Reply:
It is simply easier to write the pattern than to construct an AST explicitly. I would prefer if we can keep the pattern, since this happens at compile-time, and it doesn't change the time complexity anyway. But if you strongly think that we should switch to constructing an AST explicitly, I can do that too.

Reply:
My experience is that constructing and then parsing new syntax often leads to sub-optimal error behavior, and thus I'm always in favor of not doing it.